Instrument RPC calls to amazon SQS

stuartnelson3 opened the pull request:
Add instrumentation for SQS methods for sending/receiving/deleting messages.

Note: 
Ignoring message queues is part of the [messaging spec](https://github.com/elastic/apm/blob/master/specs/agents/tracing-instrumentation-messaging.md#ignore_message_queues-configuration), but will be included later. This is being tracked in https://github.com/elastic/apm-agent-go/issues/934.

closes https://github.com/elastic/apm-agent-go/issues/883

apmmachine commented:
## :broken_heart: Build Failed


<!-- BUILD BADGES-->
> _the below badges are clickable and redirect to their specific view in the CI or DOCS_
[![Pipeline View](https://img.shields.io/badge/pipeline-pipeline%20-green)](https://apm-ci.elastic.co/blue/organizations/jenkins/apm-agent-go%2Fapm-agent-go-mbp%2FPR-933/detail/PR-933/1//pipeline) [![Test View](https://img.shields.io/badge/test-test-green)](https://apm-ci.elastic.co/blue/organizations/jenkins/apm-agent-go%2Fapm-agent-go-mbp%2FPR-933/detail/PR-933/1//tests) [![Changes](https://img.shields.io/badge/changes-changes-green)](https://apm-ci.elastic.co/blue/organizations/jenkins/apm-agent-go%2Fapm-agent-go-mbp%2FPR-933/detail/PR-933/1//changes) [![Artifacts](https://img.shields.io/badge/artifacts-artifacts-yellow)](https://apm-ci.elastic.co/blue/organizations/jenkins/apm-agent-go%2Fapm-agent-go-mbp%2FPR-933/detail/PR-933/1//artifacts) [![preview](https://img.shields.io/badge/docs-preview-yellowgreen)](http://apm-agent-go_933.docs-preview.app.elstc.co/diff)

<!-- BUILD SUMMARY-->
<details><summary>Expand to view the summary</summary>
<p>

#### Build stats


* Build Cause: `Pull request #933 opened`



* Start Time: 2021-04-14T09:08:44.856+0000


* Duration: 9 min 53 sec


* Commit: d5b03be25089b9689f97585ee1c1f86086b2f3d0



#### Test stats :test_tube:

| Test         | Results                         |
| ------------ | :-----------------------------: |
| Failed       | 0  |
| Passed       | 2747  |
| Skipped      | 50 |
| Total        | 2797   |


#### Trends :test_tube:


![Image of Build Times](https://apm-ci.elastic.co/job/apm-agent-go/job/apm-agent-go-mbp/job/PR-933/buildTimeGraph/png)


![Image of Tests](https://apm-ci.elastic.co/job/apm-agent-go/job/apm-agent-go-mbp/job/PR-933/1/testReport/history/countGraph/png?start=0&amp;end=25)


</p>
</details>

<!-- TEST RESULTS IF ANY-->


<!-- STEPS ERRORS IF ANY -->









### Steps errors [![22](https://img.shields.io/badge/22%20-red)](https://apm-ci.elastic.co/blue/organizations/jenkins/apm-agent-go%2Fapm-agent-go-mbp%2FPR-933/detail/PR-933/1//pipeline)
  <details><summary>Expand to view the steps failures</summary>
  <p>


> Show only the first 10 steps failures

  
  
  
  
##### `Build`
<ul>
<li>Took 0 min 10 sec . View more details on <a href="https://apm-ci.elastic.co//blue/rest/organizations/jenkins/pipelines/apm-agent-go/pipelines/apm-agent-go-mbp/pipelines/PR-933/runs/1/steps/322/log/?start=0">here</a></li>
<li>Description: <code>./scripts/jenkins/build.sh</code></l1>
</ul>
  
  
  
  
##### `Build`
<ul>
<li>Took 0 min 4 sec . View more details on <a href="https://apm-ci.elastic.co//blue/rest/organizations/jenkins/pipelines/apm-agent-go/pipelines/apm-agent-go-mbp/pipelines/PR-933/runs/1/steps/424/log/?start=0">here</a></li>
<li>Description: <code>./scripts/jenkins/build.sh</code></l1>
</ul>
  
  
  
  
##### `Build`
<ul>
<li>Took 0 min 4 sec . View more details on <a href="https://apm-ci.elastic.co//blue/rest/organizations/jenkins/pipelines/apm-agent-go/pipelines/apm-agent-go-mbp/pipelines/PR-933/runs/1/steps/452/log/?start=0">here</a></li>
<li>Description: <code>./scripts/jenkins/build.sh</code></l1>
</ul>
  
  
  
  
##### `Build`
<ul>
<li>Took 0 min 10 sec . View more details on <a href="https://apm-ci.elastic.co//blue/rest/organizations/jenkins/pipelines/apm-agent-go/pipelines/apm-agent-go-mbp/pipelines/PR-933/runs/1/steps/347/log/?start=0">here</a></li>
<li>Description: <code>./scripts/jenkins/build.sh</code></l1>
</ul>
  
  
  
  
##### `Build`
<ul>
<li>Took 0 min 3 sec . View more details on <a href="https://apm-ci.elastic.co//blue/rest/organizations/jenkins/pipelines/apm-agent-go/pipelines/apm-agent-go-mbp/pipelines/PR-933/runs/1/steps/438/log/?start=0">here</a></li>
<li>Description: <code>./scripts/jenkins/build.sh</code></l1>
</ul>
  
  
  
  
##### `Build`
<ul>
<li>Took 0 min 4 sec . View more details on <a href="https://apm-ci.elastic.co//blue/rest/organizations/jenkins/pipelines/apm-agent-go/pipelines/apm-agent-go-mbp/pipelines/PR-933/runs/1/steps/467/log/?start=0">here</a></li>
<li>Description: <code>./scripts/jenkins/build.sh</code></l1>
</ul>
  
  
  
  
##### `Build`
<ul>
<li>Took 0 min 12 sec . View more details on <a href="https://apm-ci.elastic.co//blue/rest/organizations/jenkins/pipelines/apm-agent-go/pipelines/apm-agent-go-mbp/pipelines/PR-933/runs/1/steps/530/log/?start=0">here</a></li>
<li>Description: <code>./scripts/jenkins/build.sh</code></l1>
</ul>
  
  
  
  
##### `Build`
<ul>
<li>Took 0 min 4 sec . View more details on <a href="https://apm-ci.elastic.co//blue/rest/organizations/jenkins/pipelines/apm-agent-go/pipelines/apm-agent-go-mbp/pipelines/PR-933/runs/1/steps/537/log/?start=0">here</a></li>
<li>Description: <code>./scripts/jenkins/build.sh</code></l1>
</ul>
  
  
  
  
##### `Build`
<ul>
<li>Took 0 min 4 sec . View more details on <a href="https://apm-ci.elastic.co//blue/rest/organizations/jenkins/pipelines/apm-agent-go/pipelines/apm-agent-go-mbp/pipelines/PR-933/runs/1/steps/544/log/?start=0">here</a></li>
<li>Description: <code>./scripts/jenkins/build.sh</code></l1>
</ul>
  
  
  
  
##### `Error signal`
<ul>
<li>Took 0 min 0 sec . View more details on <a href="https://apm-ci.elastic.co//blue/rest/organizations/jenkins/pipelines/apm-agent-go/pipelines/apm-agent-go-mbp/pipelines/PR-933/runs/1/steps/402/log/?start=0">here</a></li>
<li>Description: <code>hudson.AbortException: script returned exit code 2</code></l1>
</ul>
  
  </p>
  </details>



<!-- LOG OUTPUT-->
### Log output

<details><summary>Expand to view the last 100 lines of log output</summary>
<p>

```
[2021-04-14T09:17:25.853Z] --- PASS: TestWrap (0.01s)
[2021-04-14T09:17:25.853Z] === RUN   TestWithContext
[2021-04-14T09:17:25.853Z] --- PASS: TestWithContext (0.00s)
[2021-04-14T09:17:25.853Z] === RUN   TestConnWithTimeout
[2021-04-14T09:17:25.853Z] --- PASS: TestConnWithTimeout (0.01s)
[2021-04-14T09:17:25.853Z] === RUN   TestWrapPipeline
[2021-04-14T09:17:25.853Z] --- PASS: TestWrapPipeline (0.00s)
[2021-04-14T09:17:25.853Z] === RUN   TestRequestContext
[2021-04-14T09:17:25.853Z] --- SKIP: TestRequestContext (0.00s)
[2021-04-14T09:17:25.853Z]     integration_test.go:119: REDIS_URL not specified
[2021-04-14T09:17:25.853Z] === RUN   TestPipelineSendReceive
[2021-04-14T09:17:25.853Z] --- SKIP: TestPipelineSendReceive (0.00s)
[2021-04-14T09:17:25.854Z]     integration_test.go:119: REDIS_URL not specified
[2021-04-14T09:17:25.854Z] === RUN   TestPipelinedTransaction
[2021-04-14T09:17:25.854Z] --- SKIP: TestPipelinedTransaction (0.00s)
[2021-04-14T09:17:25.854Z]     integration_test.go:119: REDIS_URL not specified
[2021-04-14T09:17:25.854Z] PASS
[2021-04-14T09:17:25.854Z] ok  	go.elastic.co/apm/module/apmredigo	0.034s
[2021-04-14T09:17:26.813Z] === RUN   TestMassageTemplate
[2021-04-14T09:17:26.813Z] --- PASS: TestMassageTemplate (0.00s)
[2021-04-14T09:17:26.813Z] === RUN   TestMuxMiddleware
[2021-04-14T09:17:26.813Z] --- PASS: TestMuxMiddleware (0.00s)
[2021-04-14T09:17:26.813Z] === RUN   TestInstrumentUnknownRoute
[2021-04-14T09:17:26.813Z] --- PASS: TestInstrumentUnknownRoute (0.00s)
[2021-04-14T09:17:26.813Z] === RUN   TestInstrumentWithPanicPropagation
[2021-04-14T09:17:26.813Z] === RUN   TestInstrumentWithPanicPropagation/false
[2021-04-14T09:17:26.813Z] === RUN   TestInstrumentWithPanicPropagation/true
[2021-04-14T09:17:26.813Z] --- PASS: TestInstrumentWithPanicPropagation (0.00s)
[2021-04-14T09:17:26.813Z]     --- PASS: TestInstrumentWithPanicPropagation/false (0.00s)
[2021-04-14T09:17:26.813Z]     --- PASS: TestInstrumentWithPanicPropagation/true (0.00s)
[2021-04-14T09:17:26.813Z] PASS
[2021-04-14T09:17:26.813Z] ok  	go.elastic.co/apm/module/apmgorilla	0.020s
[2021-04-14T09:17:27.770Z] === RUN   TestMiddlewareHTTPSuite
[2021-04-14T09:17:27.770Z] === RUN   TestMiddlewareHTTPSuite/TestImplicitWrite
[2021-04-14T09:17:28.037Z] === RUN   TestMiddlewareHTTPSuite/TestPanicAfterWrite
[2021-04-14T09:17:28.037Z] === RUN   TestMiddlewareHTTPSuite/TestPanicBeforeWrite
[2021-04-14T09:17:28.037Z] --- PASS: TestMiddlewareHTTPSuite (0.02s)
[2021-04-14T09:17:28.037Z]     --- PASS: TestMiddlewareHTTPSuite/TestImplicitWrite (0.00s)
[2021-04-14T09:17:28.037Z]     --- PASS: TestMiddlewareHTTPSuite/TestPanicAfterWrite (0.00s)
[2021-04-14T09:17:28.037Z]     --- PASS: TestMiddlewareHTTPSuite/TestPanicBeforeWrite (0.00s)
[2021-04-14T09:17:28.037Z] === RUN   TestEchoMiddleware
[2021-04-14T09:17:28.037Z] --- PASS: TestEchoMiddleware (0.00s)
[2021-04-14T09:17:28.037Z] === RUN   TestEchoMiddlewareUnknownRoute
[2021-04-14T09:17:28.037Z] --- PASS: TestEchoMiddlewareUnknownRoute (0.01s)
[2021-04-14T09:17:28.037Z] === RUN   TestEchoMiddlewarePanic
[2021-04-14T09:17:28.037Z] --- PASS: TestEchoMiddlewarePanic (0.01s)
[2021-04-14T09:17:28.037Z] === RUN   TestEchoMiddlewarePanicHeadersSent
[2021-04-14T09:17:28.037Z] --- PASS: TestEchoMiddlewarePanicHeadersSent (0.00s)
[2021-04-14T09:17:28.037Z] === RUN   TestEchoMiddlewareError
[2021-04-14T09:17:28.037Z] --- PASS: TestEchoMiddlewareError (0.00s)
[2021-04-14T09:17:28.037Z] PASS
[2021-04-14T09:17:28.037Z] ok  	go.elastic.co/apm/module/apmecho	0.063s
[2021-04-14T09:17:29.433Z] === RUN   TestGRPCTransport
[2021-04-14T09:17:29.433Z] --- PASS: TestGRPCTransport (0.01s)
[2021-04-14T09:17:29.433Z] === RUN   TestHTTPTransport
[2021-04-14T09:17:29.433Z] --- PASS: TestHTTPTransport (0.01s)
[2021-04-14T09:17:29.433Z] PASS
[2021-04-14T09:17:29.433Z] ok  	go.elastic.co/apm/module/apmgokit	0.028s
[2021-04-14T09:17:31.981Z] === RUN   TestMiddleware
[2021-04-14T09:17:31.981Z] --- PASS: TestMiddleware (0.01s)
[2021-04-14T09:17:31.981Z] === RUN   TestMiddlewareUnknownRoute
[2021-04-14T09:17:31.981Z] --- PASS: TestMiddlewareUnknownRoute (0.01s)
[2021-04-14T09:17:31.981Z] === RUN   TestMiddlewareControllerPanic
[2021-04-14T09:17:31.981Z] --- PASS: TestMiddlewareControllerPanic (0.01s)
[2021-04-14T09:17:31.981Z] === RUN   TestORM
[2021-04-14T09:17:31.981Z] --- PASS: TestORM (0.00s)
[2021-04-14T09:17:31.981Z] PASS
[2021-04-14T09:17:31.981Z] ok  	go.elastic.co/apm/module/apmbeego	0.042s
[2021-04-14T09:17:32.553Z] === RUN   TestGoCollector
[2021-04-14T09:17:32.553Z] --- PASS: TestGoCollector (0.01s)
[2021-04-14T09:17:32.553Z] === RUN   TestSummary
[2021-04-14T09:17:32.813Z] --- PASS: TestSummary (0.01s)
[2021-04-14T09:17:32.813Z] === RUN   TestLabels
[2021-04-14T09:17:32.813Z] --- PASS: TestLabels (0.00s)
[2021-04-14T09:17:32.813Z] PASS
[2021-04-14T09:17:32.813Z] ok  	go.elastic.co/apm/module/apmprometheus	0.031s
[2021-04-14T09:17:32.813Z] + status=0
[2021-04-14T09:17:32.813Z] + go-junit-report
[2021-04-14T09:17:32.813Z] + exit 0
[2021-04-14T09:17:32.955Z] Recording test results
[2021-04-14T09:17:33.500Z] [Checks API] No suitable checks publisher found.
[2021-04-14T09:17:35.188Z] Stage "Coverage" skipped due to earlier failure(s)
[2021-04-14T09:17:35.292Z] Stage "Benchmark" skipped due to earlier failure(s)
[2021-04-14T09:17:35.523Z] Stage "More OS" skipped due to earlier failure(s)
[2021-04-14T09:17:35.637Z] Stage "Windows" skipped due to earlier failure(s)
[2021-04-14T09:17:35.657Z] Stage "OSX" skipped due to earlier failure(s)
[2021-04-14T09:17:35.776Z] Failed in branch Windows
[2021-04-14T09:17:35.793Z] Failed in branch OSX
[2021-04-14T09:17:35.979Z] Stage "Integration Tests" skipped due to earlier failure(s)
[2021-04-14T09:17:36.083Z] Stage "Release" skipped due to earlier failure(s)
[2021-04-14T09:17:36.130Z] Stage "Release" skipped due to earlier failure(s)
[2021-04-14T09:17:36.240Z] Stage "Release" skipped due to earlier failure(s)
[2021-04-14T09:17:36.639Z] Running on Jenkins in /var/lib/jenkins/workspace/agent-go_apm-agent-go-mbp_PR-933
[2021-04-14T09:17:36.786Z] [INFO] getVaultSecret: Getting secrets
[2021-04-14T09:17:37.060Z] Masking supported pattern matches of $VAULT_ADDR or $VAULT_ROLE_ID or $VAULT_SECRET_ID
[2021-04-14T09:17:38.013Z] + chmod 755 generate-build-data.sh
[2021-04-14T09:17:38.013Z] + ./generate-build-data.sh https://apm-ci.elastic.co/blue/rest/organizations/jenkins/pipelines/apm-agent-go/apm-agent-go-mbp/PR-933/ https://apm-ci.elastic.co/blue/rest/organizations/jenkins/pipelines/apm-agent-go/apm-agent-go-mbp/PR-933/runs/1 FAILURE 532882
[2021-04-14T09:17:38.263Z] INFO: curl https://apm-ci.elastic.co/blue/rest/organizations/jenkins/pipelines/apm-agent-go/apm-agent-go-mbp/PR-933/runs/1/steps/?limit=10000 -o steps-info.json
[2021-04-14T09:17:38.814Z] INFO: curl https://apm-ci.elastic.co/blue/rest/organizations/jenkins/pipelines/apm-agent-go/apm-agent-go-mbp/PR-933/runs/1/tests/?status=FAILED -o tests-errors.json
[2021-04-14T09:17:38.814Z] INFO: curl https://apm-ci.elastic.co/blue/rest/organizations/jenkins/pipelines/apm-agent-go/apm-agent-go-mbp/PR-933/runs/1/log/ -o pipeline-log.txt

```

</p>
</details>



<!--COMMENT_GENERATED_WITH_ID_comment.id-->

axw commented on module/apmawssdkgo/sqs.go:
```diff
@@ -0,0 +1,87 @@
+func (s *apmSQS) resource() string { return s.resourceName }
+
+func (s *apmSQS) setAdditional(span *apm.Span) {
+	span.Action = s.opName
```
I think it would be good to record `context.message.queue.name`, per https://github.com/elastic/apm/blob/master/specs/agents/tracing-instrumentation-messaging.md#context-fields. Maybe add a TODO? This will require updating the model package and apm.SpanContext.

stuartnelson3 replied:
:ok_hand: adding

axw reviewed (commented):
Looks good, with the caveat that I don't really know SQS. We should be recording trace context (traceparent & tracestate) in message attributes.

stuartnelson3 commented:
I added the trace context to the message attributes, but when I was inspecting the body for `SendMessage` it seems like it's tripled, which is a bit worrisome:

```
Action=SendMessage
MessageAttribute.1.Name=attr
MessageAttribute.1.Value.DataType=String
MessageAttribute.1.Value.StringValue=string+attr
MessageAttribute.2.Name=traceContext
MessageAttribute.2.Value.DataType=String
MessageAttribute.2.Value.StringValue=00-a766982841306178091fb3f9c0c1e5b5-a49abb74fdf1b46a-01
MessageBody=msg+body
QueueUrl=https%3A%2F%2Fsqs.testing.invalid%2F123456789012%2FOtherQueue
Version=2012-11-05Action=SendMessage
MessageAttribute.1.Name=attr
MessageAttribute.1.Value.DataType=String
MessageAttribute.1.Value.StringValue=string+attr
MessageAttribute.2.Name=traceContext
MessageAttribute.2.Value.DataType=String
MessageAttribute.2.Value.StringValue=00-a766982841306178091fb3f9c0c1e5b5-a49abb74fdf1b46a-01
MessageBody=msg+body
QueueUrl=https%3A%2F%2Fsqs.testing.invalid%2F123456789012%2FOtherQueue
Version=2012-11-05Action=SendMessage
MessageAttribute.1.Name=attr
MessageAttribute.1.Value.DataType=String
MessageAttribute.1.Value.StringValue=string+attr
MessageAttribute.2.Name=traceContext
MessageAttribute.2.Value.DataType=String
MessageAttribute.2.Value.StringValue=00-a766982841306178091fb3f9c0c1e5b5-a49abb74fdf1b46a-01
MessageBody=msg+body
QueueUrl=https%3A%2F%2Fsqs.testing.invalid%2F123456789012%2FOtherQueue
Version=2012-11-05
```

I removed all the instrumentation we do (don't add `build`, `send`, `complete` to the handlers list), and it's still showing duplicates, so maybe it's a bug..?

```
Action=SendMessage
MessageAttribute.1.Name=attr
MessageAttribute.1.Value.DataType=String
MessageAttribute.1.Value.StringValue=string+attr
MessageBody=msg+body
QueueUrl=https%3A%2F%2Fsqs.testing.invalid%2F123456789012%2FOtherQueue
Version=2012-11-05Action=SendMessage
MessageAttribute.1.Name=attr
MessageAttribute.1.Value.DataType=String
MessageAttribute.1.Value.StringValue=string+attr
MessageBody=msg+body
QueueUrl=https%3A%2F%2Fsqs.testing.invalid%2F123456789012%2FOtherQueue
Version=2012-11-05Action=SendMessage
MessageAttribute.1.Name=attr
MessageAttribute.1.Value.DataType=String
MessageAttribute.1.Value.StringValue=string+attr
MessageBody=msg+body
QueueUrl=https%3A%2F%2Fsqs.testing.invalid%2F123456789012%2FOtherQueue
Version=2012-11-05Action=SendMessage
MessageAttribute.1.Name=attr
MessageAttribute.1.Value.DataType=String
MessageAttribute.1.Value.StringValue=string+attr
MessageBody=msg+body
QueueUrl=https%3A%2F%2Fsqs.testing.invalid%2F123456789012%2FOtherQueue
Version=2012-11-05
```

This is not happening in `SendMessageBatch`. I'm not sure what might have caused this,

axw commented on module/apmawssdkgo/session.go:
```diff
@@ -66,13 +72,49 @@ type service interface {
+	addMessageAttributes(req, span)
+}
+
 func send(req *request.Request) {
```
should we moving all the logic in send over to build?

stuartnelson3 replied:
Later steps construct the URL which we use to create the span name, for example (bucket, queue name, etc are pulled from the URL), and this isn't available until after the request has been built. However, we need a hook before the request is built to inject our own message attributes.

--> our build (add additional tracing data)
--> aws build (request generated)
--> our send (grab data from url)
--> aws send
--> our complete

axw replied:
thanks for the explanation. might be worth adding a comment somewhere

axw commented on module/apmawssdkgo/sqs.go:
```diff
@@ -0,0 +1,130 @@
+		if !ok {
+			return
+		}
+		input.MessageAttributes["traceContext"] = msgAttr
```
```suggestion
		input.MessageAttributes["Traceparent"] = msgAttr
```

there should be 1-3 attributes:
 - Traceparent (required)
 - Tracestate (optional, but should almost always be set)
 - Elastic-Apm-Traceparent (optional; depends on the result of the transaction's `ShouldPropagateLegacyHeader()` method)

The code should end up looking something like https://github.com/elastic/apm-agent-go/blob/0f4fcadbeb351a68c518bd88518f9903668973dd/module/apmhttp/client.go#L136-L145

For reference, here's the Ruby agent's test spec:
https://github.com/elastic/apm-agent-ruby/blob/4a3155ecebb7a189388f734d6a7478a05e80bc2f/spec/elastic_apm/spies/sqs_spec.rb#L68-L78

axw commented on module/apmawssdkgo/sqs.go:
```diff
@@ -0,0 +1,130 @@
+			return
+		}
+		for _, entry := range input.Entries {
+			entry.MessageAttributes["traceContext"] = msgAttr
```
same here

axw commented on module/apmawssdkgo/sqs_test.go:
```diff
@@ -0,0 +1,215 @@
+				w.WriteHeader(http.StatusInternalServerError)
+				return
+			}
+			io.Copy(buf, r.Body)
```
```suggestion
			buf.Reset() // reset on retry
			io.Copy(buf, r.Body)
```

axw replied:
the SDK will retry requests, I think that's why you're seeing duplicates

stuartnelson3 commented on module/apmawssdkgo/sqs_test.go:
```diff
@@ -0,0 +1,221 @@
+		if tc.hasTraceContext {
+			wrapped.Handlers.Build.PushBackNamed(request.NamedHandler{
+				Name: "spy_message_attrs_added",
+				Fn:   testTracingAttributes(t),
```
Took some inspiration from [the ruby client test](https://github.com/elastic/apm-agent-ruby/blob/4a3155ecebb7a189388f734d6a7478a05e80bc2f/spec/elastic_apm/spies/sqs_spec.rb#L68) and added another handler to inspect that the params were correctly modified.

axw replied:
nice!

axw reviewed (approved):
looks good!
